File: go/tree/binarySearchTree/bst.go
package bst

import (
	"container/list"
	"errors"
	"fmt"
)

type BSTNode struct {
	Left  *BSTNode
	Right *BSTNode
	Data  int
}

type BSTree struct {
	Root *BSTNode
}

func NewBST() *BSTree {
	return &BSTree{Root: nil}
}

func (bst *BSTree) Insert(data int) {
	bst.Root = bst.insert_internal(bst.Root, data)
}
func (bst *BSTree) insert_internal(current *BSTNode, data int) *BSTNode {
	if current == nil {
		return &BSTNode{Data: data}
	}
	if current.Data > data {
		current.Left = bst.insert_internal(current.Left, data)
	} else if current.Data < data {
		current.Right = bst.insert_internal(current.Right, data)
	}
	return current
}

// Delete
func (bst *BSTree) Delete(data int) {
	bst.delete_internal(bst.Root, data)
}

func (bst *BSTree) delete_internal(current *BSTNode, data int) *BSTNode {

	if current == nil {
		return nil
	}

	if data < current.Data {
		current.Left = bst.delete_internal(current.Left, data)
	} else if data > current.Data {
		current.Right = bst.delete_internal(current.Right, data)
	} else {
		if current.Left == nil && current.Right == nil {
			current = nil
		} else if current.Left == nil {
			current = current.Right
		} else if current.Right == nil {
			current = current.Left
		} else {

			temp := current.Right
			for temp.Left != nil {
				temp = temp.Left
			}
			current.Data = temp.Data
			current.Right = bst.delete_internal(current.Right, temp.Data)
		}
	}
	return current

}

// 查找指定值
func (bst *BSTree) Search(data int) bool {
	current := bst.Root
	for current != nil {
		if current.Data == data {
			return true
		} else if current.Data > data {
			current = current.Left
		} else {
			current = current.Right
		}
	}
	return false
}

// 最小值
func (bst *BSTree) FindMin() (int, error) {
	if bst.Root == nil {
		return -1, errors.New("tree is empty")
	}

	current := bst.Root
	for current.Left != nil {
		current = current.Left
	}
	return current.Data, nil
}

// 查找最大值
func (bst *BSTree) FindMax() (int, error) {
	if bst.Root == nil {
		return -1, fmt.Errorf("tree is empty")
	}
	current := bst.Root
	for current.Right != nil {
		current = current.Right
	}
	return current.Data, nil
}

// in-order traversal	中序遍历
func (bst *BSTree) InOrderTraversal(current *BSTNode) {
	if current != nil {
		bst.InOrderTraversal(current.Left)
		fmt.Print(current.Data, ",")
		bst.InOrderTraversal(current.Right)
	}
}

// pre order traversal 前序遍历
func (bst *BSTree) PreOrderTraversal(current *BSTNode) {
	if current != nil {
		fmt.Print(current.Data, ",")
		bst.PreOrderTraversal(current.Left)
		bst.PreOrderTraversal(current.Right)
	}
}

// post order traversal   left -> right -> root  后续遍历
func (bst *BSTree) PostOrderTraversal(current *BSTNode) {
	if current != nil {
		bst.PostOrderTraversal(current.Left)
		bst.PostOrderTraversal(current.Right)
		fmt.Print(current.Data, ",")
	}
}

/**
*  level order traversal
* 层序遍历
 */
func (bst *BSTree) LevelOrderTraversal() {
	l := list.New()
	l.PushBack(bst.Root)
	for l.Len() > 0 {
		e := l.Front()
		l.Remove(e)
		node := e.Value.(*BSTNode)
		fmt.Print(node.Data, ",")
		if node.Left != nil {
			l.PushBack(node.Left)
		}
		if node.Right != nil {
			l.PushBack(node.Right)
		}
	}
}

func (bst *BSTree) LevelOrderToList() []int {
	arr := make([]int, 0)
	l := list.New()
	l.PushBack(bst.Root)
	for l.Len() > 0 {
		e := l.Front()
		l.Remove(e)
		node := e.Value.(*BSTNode)
		arr = append(arr, node.Data)
		if node.Left != nil {
			l.PushBack(node.Left)
		}
		if node.Right != nil {
			l.PushBack(node.Right)
		}
	}
	return arr
}

File: dart/src/tree/binarySearchTree/bst.dart
import 'dart:collection';

class BSTNode<T extends Comparable> {
  T data;
  BSTNode<T>? left;
  BSTNode<T>? right;

  BSTNode(this.data, this.left, this.right);
}

class BinarySearchTree<T extends Comparable> {
  BSTNode<T>? root;
  BinarySearchTree() {
    this.root = null;
  }

  void insert(T data) {
    this.root = insertNode(this.root, data);
  }

  BSTNode<T> insertNode(BSTNode<T>? node, T data) {
    if (node == null) {
      return BSTNode(data, null, null);
    }
    if (data.compareTo(node.data) < 0) {
      node.left = insertNode(node.left, data);
    } else if (data.compareTo(node.data) > 0) {
      node.right = insertNode(node.right, data);
    }
    return node;
  }

  void delete(T data) {
    this.root = deleteNode(this.root, data);
  }

  BSTNode<T>? deleteNode(BSTNode<T>? root, T data) {
    if (root == null) {
      return null;
    }
    if (data.compareTo(root.data) < 0) {
      root.left = deleteNode(root.left, data);
    }
    if (data.compareTo(root.data) > 0) {
      root.right = deleteNode(root.right, data);
    }
    if (data.compareTo(root.data) == 0) {
      if (root.left == null && root.right == null) {
        root = null;
      } else if (root.left == null) {
        root = root.right;
      } else if (root.right == null) {
        root = root.left;
      } else {
        var node = root.right;
        while (node?.left != null) {
          node = node?.left;
        }
        root.data = node!.data;
        root.right = deleteNode(root.right, node.data);
      }
    }
    return root;
  }

  bool search(T data) {
    BSTNode<T>? current = this.root;
    while (current != null) {
      if (data.compareTo(current.data) < 0) {
        current = current.left;
      } else if (data.compareTo(current.data) > 0) {
        current = current.right;
      } else {
        return true;
      }
    }
    return false;
  }

  T findMin() {
    BSTNode<T>? current = this.root;
    while (current?.left != null) {
      current = current?.left;
    }
    return current!.data;
  }

  T findMax() {
    BSTNode<T>? current = this.root;
    while (current?.right != null) {
      current = current?.right;
    }
    return current!.data;
  }

  int height(BSTNode<T>? node) {
    if (node == null) {
      return -1;
    }
    int leftHeight = height(node.left);
    int rightHeight = height(node.right);
    return leftHeight > rightHeight ? leftHeight + 1 : rightHeight + 1;
  }

  void inOrder(BSTNode<T>? node) {
    if (node == null) {
      return;
    }
    inOrder(node.left);
    print(node.data);
    inOrder(node.right);
  }

  void preOrder(BSTNode<T>? node) {
    if (node == null) {
      return;
    }
    print(node.data);
    preOrder(node.left);
    preOrder(node.right);
  }

  void postOrder(BSTNode<T>? node) {
    if (node == null) {
      return;
    }
    postOrder(node.left);
    postOrder(node.right);
    print(node.data);
  }

  void levelOrder(BSTNode<T>? node) {
    if (node == null) {
      return;
    }
    Queue<BSTNode<T>> queue = Queue();
    queue.addLast(node);
    while (!queue.isEmpty) {
      BSTNode<T> current = queue.removeFirst();
      print(current.data);
      if (current.left != null) {
        queue.addLast(current.left!);
      }
      if (current.right != null) {
        queue.addLast(current.right!);
      }
    }
  }

  List toList() {
    List list = [];
    if (root != null) {
      Queue<BSTNode<T>> queue = Queue();
      queue.addLast(root!);
      while (!queue.isEmpty) {
        BSTNode<T> current = queue.removeFirst();
        list.add(current.data);
        if (current.left != null) {
          queue.addLast(current.left!);
        }
        if (current.right != null) {
          queue.addLast(current.right!);
        }
      }
    }
    return list;
  }

  get isEmpty {
    return this.root == null;
  }

//   void insert(T data) {
//     BSTNode<T> newNode = BSTNode(data, null, null);
//     if (this.root == null) {
//       this.root = newNode;
//     } else {
//       BSTNode<T>? current = this.root;
//       BSTNode<T>? parent;
//       while (true) {
//         parent = current;
//         if (data.compareTo(current?.data) < 0) {
//           current = current?.left;
//           if (current == null) {
//             parent?.left = newNode;
//             break;
//           }
//         } else {
//           current = current?.right;
//           if (current == null) {
//             parent?.right = newNode;
//             break;
//           }
//         }
//       }
//     }
//   }
}
